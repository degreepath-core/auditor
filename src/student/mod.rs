use crate::audit::Transcript;
use overrides::Override;
use serde::{Deserialize, Serialize};

mod data;
pub use data::{AreaDescriptor, AttendanceInstance, CourseInstance, OrganizationDescriptor, PerformanceInstance};
pub use data::{CourseType, GradeOption, Semester};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct StudentData {
	pub transcript: Transcript,
	pub organizations: Vec<OrganizationDescriptor>,
	pub performances: Vec<PerformanceInstance>,
	pub attendances: Vec<AttendanceInstance>,
	pub areas: Vec<AreaDescriptor>,
	pub overrides: Vec<Override>,
}

impl StudentData {
	pub fn update_transcript(self, courses: &[CourseInstance]) -> StudentData {
		StudentData {
			transcript: Transcript::new(courses),
			..self
		}
	}
}

pub mod overrides {
	use super::{AreaDescriptor, AttendanceInstance, CourseInstance, PerformanceInstance};
	use serde::{Deserialize, Serialize};

	#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
	pub struct Override {
		pub path: Vec<PathSegment>,
		#[serde(default)]
		pub value: Option<Value>,
		pub mode: Mode,
	}

	#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
	pub enum Mode {
		Pass,
		AllowValue,
	}

	#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
	pub enum Value {
		Course(CourseInstance),
		Area(AreaDescriptor),
		Performance(PerformanceInstance),
		Attendance(AttendanceInstance),
	}

	#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
	pub enum PathSegment {
		Root,
		Requirement(String),
		Path(String),
		Index(u16),
	}

	mod tests {
		// /, :Core, .result, .of, #4, .either, #0
		// [Root, Requirement("Core"), Path("result"), Path("of"), Index(4), Path("either"), Index(0)]
	}
}
