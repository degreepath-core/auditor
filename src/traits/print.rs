use std::fmt;

pub type Result = std::result::Result<String, fmt::Error>;

pub trait Print {
	fn print(&self) -> Result;

	fn print_indented(&self, level: usize) -> Result {
		let printed = self.print()?;

		if printed.contains('\n') {
			Ok(indent_non_empty_lines(&printed, level))
		} else {
			Ok(printed.to_string())
		}
	}
}

fn indent_non_empty_lines(s: &str, times: usize) -> String {
	let lines: Vec<_> = s.lines().collect();

	if let Some((first, rest)) = lines.split_first() {
		let indent = "    ".repeat(times);

		// don't indent empty lines
		let rest = rest.iter().map(|l| {
			if l.trim() != "" {
				format!("{}{}", indent, l)
			} else {
				String::from("")
			}
		});

		// skip the first blank line
		let rest = rest.skip_while(|l| l.is_empty());

		let rest = rest.collect::<Vec<_>>().join("\n");

		format!("{}\n{}", first, rest)
	} else {
		s.to_string()
	}
}
