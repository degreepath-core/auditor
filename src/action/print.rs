use super::{Action, Command, Operator, Value};
use crate::traits::print;
use std::fmt::Write;

impl print::Print for Action {
	fn print(&self) -> print::Result {
		let mut output = String::new();

		match (&self.lhs, &self.op, &self.rhs) {
			(Value::String(s), Some(op), Some(val)) if *s == Command::Count => {
				write!(&mut output, "{} {}", op.print()?, val.print()?)?
			}
			(Value::String(s), Some(op), Some(val)) if *s == Command::Sum => {
				write!(&mut output, "{} {}", op.print()?, val.print()?)?
			}
			(Value::String(s), Some(Operator::GreaterThan), Some(val)) if *s == Command::Average => {
				write!(&mut output, "above {}", val.print()?)?
			}
			(Value::String(s), Some(Operator::GreaterThanEqualTo), Some(val)) if *s == Command::Average => {
				write!(&mut output, "at or above {}", val.print()?)?
			}
			_ => unimplemented!(
				"in Action's printer, the combo of `{:?}`, `{:?}`, and `{:?}`",
				&self.lhs,
				&self.op,
				&self.rhs
			),
		};

		Ok(output)
	}
}
