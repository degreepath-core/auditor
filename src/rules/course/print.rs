use super::*;
use crate::traits::print;
use crate::util;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;
		let mut output = String::new();

		write!(&mut output, "{}", self.course)?;

		if let Some(section) = &self.section {
			write!(&mut output, "{}", section)?;
		}

		let mut annotations = match (self.year, &self.semester) {
			(Some(year), None) => Some(util::expand_year(year, "dual")),
			(None, Some(semester)) => Some(semester.to_string()),
			(Some(year), Some(semester)) => Some(format!("{} {}", semester, util::expand_year(year, "short"))),
			(None, None) => None,
		};

		if let Some(lab) = self.lab {
			if lab {
				match &annotations {
					Some(ann) => {
						annotations = Some(format!("Lab; {}", ann));
					}
					None => {
						annotations = Some("Lab".to_string());
					}
				}
			}
		}

		match (&self.grade, &annotations) {
			(Some(grade), Some(ann)) => annotations = Some(format!("{}; {} or higher", ann, String::from(grade))),
			(Some(grade), None) => annotations = Some(format!("{} or higher", String::from(grade))),
			_ => (),
		}

		if let Some(annotations) = annotations {
			write!(&mut output, " ({})", annotations)?;
		}

		Ok(output)
	}
}
