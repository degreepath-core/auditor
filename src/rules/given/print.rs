use super::{
	CourseRule, Given, GivenAreasWhatOptions, GivenAttendancesWhatOptions, GivenCoursesWhatOptions,
	GivenPerformancesWhatOptions, RepeatMode, Rule,
};
use crate::action;
use crate::filter::{AreaClause, AttendanceClause, CourseClause, PerformanceClause};
use crate::limit::Limiter;
use crate::traits::print::{self, Print};
use crate::util::Oxford;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;

		let mut output = String::new();

		let rule = match &self.given {
			Given::AllCourses { what, filter, limit } => self.print_given_all_courses(what, filter, limit)?,
			Given::TheseCourses {
				what,
				courses,
				repeats: mode,
				filter,
				limit,
			} => self.print_given_these_courses(courses, mode, what, filter, limit)?,
			Given::TheseRequirements {
				what,
				requirements,
				filter,
				limit,
			} => self.print_given_these_requirements(requirements, what, filter, limit)?,
			Given::NamedVariable {
				save,
				what,
				filter,
				limit,
			} => self.print_given_save(save, what, filter, limit)?,
			Given::Areas { what, filter } => self.print_given_areas(what, filter)?,
			Given::Performances { what, filter } => self.print_given_performances(what, filter)?,
			Given::Attendances { what, filter } => self.print_given_attendances(what, filter)?,
		};

		write!(&mut output, "{}", rule)?;

		Ok(output)
	}
}

impl Rule {
	fn print_given_all_courses(
		&self,
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();
		let action = self.action.print()?;
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};
		let limits = if let Some(limits) = &limit {
			let stringified = limits
				.iter()
				.map(|l| format!("at most {} {}", l.at_most, l.filter.print().unwrap()))
				.collect::<Vec<_>>()
				.oxford("and");
			format!(" {}", stringified)
		} else {
			"".to_owned()
		};

		match &what {
			What::Courses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };

				write!(&mut output, "take {} {}{}{}", action, word, filter, limits)?;
			}
			What::DistinctCourses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "distinct courses" } else { "course" };

				write!(&mut output, "take {} {}{}", action, word, filter)?;
			}
			What::Credits => {
				let plur = self.action.should_pluralize();
				let word = if plur { "credits" } else { "credit" };
				let filter = if filter.is_empty() {
					"".to_owned()
				} else {
					format!(" taken{}", filter)
				};

				write!(
					&mut output,
					"have enough courses{} to obtain {} {}",
					filter, action, word
				)?;
			}
			What::Departments => {
				let plur = self.action.should_pluralize();
				let word = if plur { "departments" } else { "department" };
				let filter = if filter.is_empty() {
					"".to_owned()
				} else {
					format!(" taken{}", filter)
				};

				write!(&mut output, "have enough courses{} to span {} {}", filter, action, word)?;
			}
			What::Grades => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };
				let filter = if filter.is_empty() {
					"".to_owned()
				} else {
					format!(" taken{}", filter)
				};

				write!(
					&mut output,
					"maintain an average GPA {} from {}{}",
					action, word, filter
				)?;
			}
			What::Terms => {
				let plur = self.action.should_pluralize();
				let word = if plur { "terms" } else { "term" };

				write!(
					&mut output,
					"have taken enough courses{} to span {} {}",
					filter, action, word
				)?;
			}
		}

		Ok(output)
	}

	fn print_given_areas(&self, what: &GivenAreasWhatOptions, filter: &Option<AreaClause>) -> print::Result {
		use std::fmt::Write;
		use GivenAreasWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Areas => {
				// TODO: find a better way to special-case "exactly one" major
				let action = self.action.print()?;
				let action = action.replace("exactly ", "");
				write!(&mut output, "declare {}{}", action, filter)?;
			}
		}

		Ok(output)
	}

	fn print_given_performances(
		&self,
		what: &GivenPerformancesWhatOptions,
		filter: &Option<PerformanceClause>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenPerformancesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Performances => {
				let action = self.action.print()?;
				write!(&mut output, "perform {} recitals{}", action, filter)?;
			}
		}

		Ok(output)
	}

	fn print_given_attendances(
		&self,
		what: &GivenAttendancesWhatOptions,
		filter: &Option<AttendanceClause>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenAttendancesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Attendances => {
				let action = self.action.print()?;
				write!(&mut output, "attend {}{} recitals", action, filter)?;
			}
		}

		Ok(output)
	}

	fn print_given_these_courses(
		&self,
		courses: &[CourseRule],
		mode: &RepeatMode,
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		_limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => Some(format!(" {}", f.print()?)),
			None => None,
		};

		let courses: Vec<String> = courses.iter().map(|r| r.print().unwrap()).collect();

		match (mode, what) {
			(RepeatMode::First, What::Courses) | (RepeatMode::Last, What::Courses) => {
				match courses.len() {
					1 => {
						// TODO: expose last vs. first in output somehow?
						write!(&mut output, "take {}", courses.oxford("and"))?;
					}
					2 => match (&self.action.lhs, &self.action.op, &self.action.rhs) {
						(
							action::Command::Count,
							Some(action::Operator::GreaterThanEqualTo),
							Some(action::Value::Integer(n)),
						) => match n {
							1 => {
								write!(&mut output, "take either {} or {}", courses[0], courses[1])?;
							}
							2 => {
								write!(&mut output, "take both {} and {}", courses[0], courses[1])?;
							}
							_ => panic!("should not require <1 or >len of the number of courses given"),
						},
						_ => unimplemented!("most actions on two-up given:these-courses rules"),
					},
					3...5 => {
						// TODO: expose last vs. first in output somehow?
						let plur = self.action.should_pluralize();
						let word = if plur { "courses" } else { "course" };
						write!(
							&mut output,
							"take {} {} from among {}",
							self.action.print()?,
							word,
							courses.oxford("and")
						)?;
					}
					_ => {
						// TODO: expose last vs. first in output somehow?
						let plur = self.action.should_pluralize();
						let word = if plur { "courses" } else { "course" };

						let as_list: Vec<_> = courses.iter().map(|l| format!("- {}", l)).collect();

						write!(
							&mut output,
							"take {} {} from among the following:\n\n{}",
							self.action.print()?,
							word,
							as_list.join("\n")
						)?;
					}
				}
			}
			(RepeatMode::All, What::Courses) => {
				// TODO: special-case "once" and "twice"
				let plur = self.action.should_pluralize();
				let word = if plur { "times" } else { "time" };

				match (&self.action.lhs, &self.action.op, &self.action.rhs) {
					(
						action::Command::Count,
						Some(action::Operator::GreaterThanEqualTo),
						Some(action::Value::Integer(1)),
					) => match courses.len() {
						1...5 => {
							write!(
								&mut output,
								"take {} {} {}",
								courses.oxford("or"),
								self.action.print()?,
								word
							)?;
						}
						_ => {
							let as_list: Vec<_> = courses.iter().map(|l| format!("- {}", l)).collect();

							write!(
								&mut output,
								"take {} of the following courses:\n\n{}",
								self.action.print()?,
								as_list.join("\n")
							)?;
						}
					},
					_ => match courses.len() {
						1 => {
							write!(
								&mut output,
								"take {} {} {}",
								courses.oxford("and"),
								self.action.print()?,
								word
							)?;
						}
						_ => {
							write!(
								&mut output,
								"take a combination of {} {} {}",
								courses.oxford("and"),
								self.action.print()?,
								word
							)?;
						}
					},
				}
			}
			(RepeatMode::All, What::Credits) => {
				// TODO: special-case "once" and "twice"
				let plur = self.action.should_pluralize();
				let word = if plur { "credits" } else { "credit" };

				write!(
					&mut output,
					"take {} enough times to yield {} {}",
					courses.oxford("and"),
					self.action.print()?,
					word
				)?;
			}
			(RepeatMode::All, What::Terms) => {
				// TODO: special-case "once" and "twice"
				let plur = self.action.should_pluralize();
				let word = if plur { "terms" } else { "term" };

				write!(
					&mut output,
					"take {} enough times to span {} {}",
					courses.oxford("and"),
					self.action.print()?,
					word
				)?;
			}
			_ => unimplemented!("certain modes of given:these-courses"),
		}

		if let Some(f) = filter {
			write!(&mut output, "{}", f)?;
		}

		Ok(output)
	}

	fn print_given_these_requirements(
		&self,
		requirements: &[String],
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		_limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();

		let requirements: Vec<String> = requirements.iter().map(|r| format!("“{}”", r)).collect();

		writeln!(&mut output, "have the following be true:\n")?;
		let mut index = 0;

		index += 1;
		match requirements.len() {
			0 => panic!("no requirements given!"),
			1 | 2 | 3 => {
				let singular = requirements.len() == 1;
				let word = if singular { "requirement" } else { "requirements" };
				writeln!(
					&mut output,
					"{index}. given the results of the {list} {word},",
					index = index,
					list = requirements.oxford("and"),
					word = word
				)?;
			}
			_ => {
				writeln!(
					&mut output,
					"{index}. given the results of the following requirements",
					index = index
				)?;
				for req in requirements {
					writeln!(&mut output, "    - {}", req)?;
				}
			}
		};

		match &filter {
			Some(f) => {
				index += 1;
				writeln!(
					&mut output,
					"{index}. restricted to only courses taken {filter},",
					index = index,
					filter = f.print()?
				)?;
			}
			None => (),
		};

		index += 1;
		let pluralize = self.action.should_pluralize();

		match &what {
			What::Courses => {
				let word = if pluralize { "courses" } else { "course" };

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::DistinctCourses => {
				let word = if pluralize {
					"distinct courses"
				} else {
					"distinct course"
				};

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::Credits => {
				let word = if pluralize { "credits" } else { "credit" };

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::Departments => {
				let word = if pluralize {
					"distinct departments"
				} else {
					"department"
				};

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::Grades => {
				writeln!(
					&mut output,
					"{index}. there must be an average GPA {action}",
					index = index,
					action = self.action.print()?,
				)?;
			}
			What::Terms => {
				let word = if pluralize { "terms" } else { "term" };

				writeln!(
					&mut output,
					"{index}. there must be courses in {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
		};

		Ok(output)
	}

	fn print_given_save(
		&self,
		save: &str,
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		_limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" taken {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Courses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(&mut output, "there must be {} {}{}", self.action.print()?, word, filter)?;
				// write!(&mut output, " in the subset “{}”", save)?;
			}
			What::DistinctCourses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "distinct courses" } else { "course" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(&mut output, "there must be {} {}{}", self.action.print()?, word, filter)?;
				// write!(&mut output, " in the subset “{}”", save)?;
			}
			What::Credits => {
				let plur = self.action.should_pluralize();
				let word = if plur { "credits" } else { "credit" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(
					&mut output,
					"there must be enough courses{} to obtain {} {}",
					filter,
					self.action.print()?,
					word
				)?;
			}
			What::Departments => {
				let plur = self.action.should_pluralize();
				let word = if plur { "departments" } else { "department" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(
					&mut output,
					"there must be enough courses{} to span {} {}",
					filter,
					self.action.print()?,
					word
				)?;
			}
			What::Grades => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };

				write!(&mut output, "courses from the subset “{}” ", save)?;
				write!(
					&mut output,
					"must maintain an average GPA {} from {}{}",
					self.action.print()?,
					word,
					filter
				)?;
			}
			What::Terms => {
				let plur = self.action.should_pluralize();
				let word = if plur { "terms" } else { "term" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(
					&mut output,
					"there must be enough courses{} to span {} {}",
					filter,
					self.action.print()?,
					word
				)?;
			}
		}

		Ok(output)
	}
}
