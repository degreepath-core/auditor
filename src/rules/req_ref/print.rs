use super::*;
use crate::traits::print;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;
		let mut output = String::new();

		write!(&mut output, "“{}”", self.name)?;

		if self.optional {
			write!(&mut output, " (optional)")?;
		}

		Ok(output)
	}
}
