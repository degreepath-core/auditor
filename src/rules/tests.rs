use super::*;
use crate::student::Semester;
use crate::traits::print::Print;
use insta::{assert_debug_snapshot_matches, assert_yaml_snapshot_matches};
use pretty_assertions::assert_eq;

#[test]
fn deserialize_simple_course_in_array() {
	let data = "---
- STAT 214";

	let deserialized: Vec<Rule> = serde_yaml::from_str(&data).unwrap();

	assert_debug_snapshot_matches!(deserialized, @r###"[
    Course(
        Rule {
            course: "STAT 214",
            grade: None,
            section: None,
            year: None,
            semester: None,
            lab: None,
            can_match_used: None
        }
    )
]"###);
}

#[test]
fn serialize() {
	let course_a = course::Rule {
		course: "ASIAN 101".to_string(),
		..Default::default()
	};
	let course_b = course::Rule {
		course: "ASIAN 101".to_string(),
		semester: Some(Semester::Fall),
		year: Some(2014),
		..Default::default()
	};
	let data = vec![
		Rule::Course(course_a.clone()),
		Rule::Course(course_b.clone()),
		Rule::Requirement(req_ref::Rule {
			name: "Name".to_string(),
			optional: true,
		}),
		Rule::CountOf(count_of::Rule {
			count: count_of::Counter::Number(1),
			of: vec![Rule::Course(course_a.clone())],
		}),
		Rule::Both(both::Rule {
			both: (
				Box::new(Rule::Course(course_a.clone())),
				Box::new(Rule::Course(course_b.clone())),
			),
		}),
		Rule::Either(either::Rule {
			either: (
				Box::new(Rule::Course(course_a.clone())),
				Box::new(Rule::Course(course_b.clone())),
			),
		}),
		Rule::Given(given::Rule {
			given: given::Given::AllCourses {
				what: given::GivenCoursesWhatOptions::Courses,
				filter: Some(crate::filter::CourseClause::default()),
				limit: Some(vec![]),
			},
			action: "count < 2".parse().unwrap(),
		}),
		Rule::Do(action_only::Rule {
			action: crate::action::LhsValueAction {
				lhs: crate::action::Value::String("a".to_string()),
				op: Some(crate::action::Operator::LessThan),
				rhs: Some(crate::action::Value::String("b".to_string())),
			},
		}),
	];

	assert_yaml_snapshot_matches!("serialize_mix_1", data);
}

#[test]
fn deserialize() {
	let data = r#"---
- {type: course, course: ASIAN 101}
- type: course
  course: ASIAN 101
  section: ~
  year: 2014
  semester: Fall
  lab: ~
- type: requirement
  name: Name
  optional: true
- type: count-of
  count: 1
  of:
    - {type: course, course: ASIAN 101}
- type: both
  both:
    - {type: course, course: ASIAN 101}
    - type: course
      course: ASIAN 101
      section: ~
      year: 2014
      semester: Fall
      lab: ~
- type: either
  either:
    - {type: course, course: ASIAN 101}
    - type: course
      course: ASIAN 101
      section: ~
      year: 2014
      semester: Fall
      lab: ~
- type: given
  given: courses
  what: courses
  where: {}
  limit: []
  do: count < 2
- {type: do, do: {lhs: a, op: <, rhs: b}}"#;

	let actual: Vec<Rule> = serde_yaml::from_str(&data).unwrap();
	assert_debug_snapshot_matches!("deserialize_mix", actual);
}

#[test]
fn deserialize_shorthands() {
	let data = r#"---
- ASIAN 101
- {type: course, course: ASIAN 101}
- {type: course, course: ASIAN 102, year: 2014, semester: Fall}
- {type: requirement, name: Name 1}
- {type: requirement, name: Name 2, optional: true}
- type: count-of
  count: 1
  of:
    - ASIAN 101
- type: both
  both:
    - ASIAN 101
    - {type: course, course: ASIAN 102, year: 2014, semester: Fall}
- type: either
  either:
    - ASIAN 101
    - {type: course, course: ASIAN 102, year: 2014, semester: Fall}
- type: given
  given: courses
  what: courses
  do: count < 2
- {type: do, do: {lhs: a, op: <, rhs: b}}"#;

	let actual: Vec<Rule> = serde_yaml::from_str(&data).unwrap();
	assert_debug_snapshot_matches!("deserialize_shorthands", actual);
}

#[test]
fn pretty_print() {
	let data = r#"---
- ASIAN 101
- {type: course, course: ASIAN 101}
- {type: course, course: ASIAN 102, year: 2014, semester: Fall}
- {type: requirement, name: Name 1}
- {type: requirement, name: Name 2, optional: true}
- {type: count-of, count: 1, of: [ASIAN 101]}
- {type: both, both: [ASIAN 101, {type: course, course: ASIAN 102, year: 2014, semester: Fall}]}
- {type: either, either: [ASIAN 101, {type: course, course: ASIAN 102, year: 2014, semester: Fall}]}
- {type: given, given: courses, what: courses, do: count < 2}
- {type: do, do: {lhs: X, op: <, rhs: Y}}
"#;

	let actual: Vec<Rule> = serde_yaml::from_str(&data).unwrap();
	let actual: Vec<String> = actual
		.into_iter()
		.filter_map(|r| r.print().ok())
		.collect();

	assert_debug_snapshot_matches!("pretty_print_mix", actual);
}

#[test]
fn dance_seminar() {
	let expected = Rule::Given(given::Rule {
		given: given::Given::NamedVariable {
			save: "Senior Dance Seminars".to_string(),
			what: given::GivenCoursesWhatOptions::Courses,
			filter: None,
			limit: None,
		},
		action: "count >= 1".parse().unwrap(),
	});

	let data = r#"---
type: given
given: save
save: "Senior Dance Seminars"
what: courses
do: count >= 1
"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_eq!(actual, expected);

	let data = r#"---
type: given
given: save
save: "Senior Dance Seminars"
what: courses
do:
  lhs: Count
  op: GreaterThanEqualTo
  rhs:
    Integer: 1
"#;

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_eq!(actual, expected);
}
