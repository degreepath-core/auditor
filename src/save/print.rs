use super::SaveBlock;
use crate::rules::given::GivenForSaveBlock as Given;
use crate::traits::print;

impl print::Print for SaveBlock {
	fn print(&self) -> print::Result {
		use std::fmt::Write;

		let mut output = String::new();

		match &self.given {
			Given::AllCourses { .. } => match &self.filter {
				Some(f) => {
					write!(&mut output, "Given the subset of courses from your transcript, limited to only courses taken {}, as “{}”:\n\n", f.print()?, self.name)?;

					writeln!(&mut output, "| “{}” |", self.name)?;
					writeln!(&mut output, "| {} |", "-".repeat(self.name.chars().count() + 2))?;
					writeln!(&mut output, "| (todo: list matching courses here) |")?;
				}
				None => {
					write!(
						&mut output,
						"Given the courses from your transcript as “{}”:\n\n",
						self.name
					)?;
					writeln!(&mut output, "| “{}” |", self.name)?;
					writeln!(&mut output, "| {} |", "-".repeat(self.name.chars().count() + 2))?;
					writeln!(&mut output, "| (todo: list all??? courses here???) |")?;
				}
			},
			Given::TheseCourses { courses, .. } => {
				write!(&mut output, "Given the intersection between this set of courses and the courses from your transcript, as “{}”:\n\n", self.name)?;

				let courses = courses.iter().map(|r| r.print().unwrap()).collect::<Vec<String>>();

				let transcript = "Transcript";
				let name_width = self.name.chars().count();

				writeln!(&mut output, "| “{}” | {} |", self.name, transcript)?;
				writeln!(
					&mut output,
					"| {} | {} |",
					"-".repeat(name_width + 2),
					"-".repeat(transcript.len()),
				)?;
				for c in courses {
					writeln!(&mut output, "| {} | (todo: fill out if match) |", c)?;
				}
			}
			Given::TheseRequirements { requirements, .. } => {
				use crate::util::Oxford;

				let req_names = requirements
					.iter()
					.map(|r| format!("“{}”", r))
					.collect::<Vec<String>>()
					.oxford("and");

				write!(
					&mut output,
					"Given the courses which fulfilled the {} requirements",
					req_names
				)?;

				if let Some(filter) = &self.filter {
					write!(&mut output, ", limited to only courses taken {}", filter.print()?)?;
				};

				writeln!(&mut output, ", as “{}”:", self.name)?;

				writeln!(&mut output)?;

				writeln!(&mut output, "| “{}” |", self.name)?;
				writeln!(&mut output, "| {} |", "-".repeat(self.name.chars().count() + 2))?;
				// todo: list matched courses from the referenced save here
				writeln!(&mut output, "| (todo: list matched courses here) |")?;

				writeln!(&mut output)?;

				if let Some(_action) = &self.action {
					// todo: describe what the save's action is doing
					writeln!(&mut output, "> todo: describe what the save's action is doing")?;
				}

				if self.limit.is_some() {
					// todo: describe what the save's limiters do
					writeln!(&mut output, "> todo: describe what the save's limiters do")?;
				}

				// todo: describe what the save will generate
				writeln!(&mut output, "> todo: describe what the save will generate")?;
			}
			Given::NamedVariable { save, .. } => match &self.filter {
				Some(f) => {
					write!(
						&mut output,
						"Given the subset named “{}”, limited it to only courses taken {}, as “{}”:\n\n",
						save,
						f.print()?,
						self.name
					)?;

					writeln!(&mut output, "| “{}” |", self.name)?;
					writeln!(&mut output, "| {} |", "-".repeat(self.name.chars().count() + 2))?;
					writeln!(&mut output, "| (todo: list matching courses here) |")?;
				}
				None => {
					write!(
						&mut output,
						"Given the subset named “{}”, as “{}”:\n\n",
						save, self.name
					)?;
					writeln!(&mut output, "| “{}” |", self.name)?;
					writeln!(&mut output, "| {} |", "-".repeat(self.name.chars().count() + 2))?;
					writeln!(&mut output, "| (todo: list all??? courses here???) |")?;
				}
			},
		}

		// TODO: add a special case for do-block-having SaveBlocks, to show the computed value
		// TODO: decide on something to do for given:these-courses, repeats:all/first/last
		// TODO: add limiter output

		if self.limit.is_some() {
			writeln!(&mut output, "\ntodo: there's a limiter")?;
		}

		Ok(output)
	}
}
