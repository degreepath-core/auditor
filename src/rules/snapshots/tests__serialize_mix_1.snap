---
created: "2019-03-03T03:16:53.929766Z"
creator: insta@0.6.3
source: src/rules/tests.rs
expression: data
---
- course: ASIAN 101
- course: ASIAN 101
  grade: ~
  section: ~
  year: 2014
  semester: Fall
  lab: ~
  can_match_used: ~
- requirement: Name
  optional: true
- count: 1
  of:
    - course: ASIAN 101
- both:
    - course: ASIAN 101
    - course: ASIAN 101
      grade: ~
      section: ~
      year: 2014
      semester: Fall
      lab: ~
      can_match_used: ~
- either:
    - course: ASIAN 101
    - course: ASIAN 101
      grade: ~
      section: ~
      year: 2014
      semester: Fall
      lab: ~
      can_match_used: ~
- given: courses
  what: courses
  where:
    attribute: ~
    course: ~
    credits: ~
    gereqs: ~
    graded: ~
    institution: ~
    level: ~
    number: ~
    type: ~
    semester: ~
    subject: ~
    year: ~
  limit: []
  do:
    lhs: Count
    op: LessThan
    rhs:
      Integer: 2
- do:
    lhs:
      String: a
    op: LessThan
    rhs:
      String: b
